### ISSUE TITLE
Slower string concatenation in CPython 3.11

### ISSUE THREAD

**Issue by zephyr111** [python/cpython#99862]
Hello,

We have found a regression between CPython 3.10.8 and CPython 3.11 resulting in string concatenation to be significantly slower in loops on Windows 10. This is described in details in this [StackOverflow post](https://stackoverflow.com/questions/74605279/python-3-11-worse-optimized-than-3-10/74607850).

Here is a minimal, reproducible example of benchmarking code:

```python
import time
a = 'a'

start = time.time()
for _ in range(1000000):
    a += 'a'
end = time.time()

print(a[:5], (end-start) * 1000)
```

CPython 3.11.0 is about 100 times slower than CPython 3.10.8 due to a quadratic running time (as opposed to a linear running time for CPython 3.10.8).

The analysis shows that CPython 3.10.8 was generating an INPLACE_ADD instruction so `PyUnicode_Append` is called at runtime, while CPython 3.11.0 new generates a BINARY_OP instruction so `PyUnicode_Concat` is actually called. The later function creates a new bigger string reducing drastically the performance of the string appending loop in the provided code. This appears to be related to the issue #89799 . I think if we want to replace INPLACE_ADD with a BINARY_OP, then an optimization checking the number of references (so to eventually do an in-place operation) is missing in the code of CPython 3.11.0. What do you think about it?

My environment is an embedded CPython 3.10.8 and an embedded CPython 3.11.0, both running on Windows 10 (22H2) with a x86-64 processor (i5-9600KF).

**Comment by ned-deily** [comment]
@brandtbucher @pablogsal

**Comment by sweeneyde** [comment]
Python 3.11 has a specialized (a la [PEP 659](https://peps.python.org/pep-0659/)) instruction called `BINARY_OP_INPLACE_ADD_UNICODE`:

https://github.com/python/cpython/blob/5bbf8ed8fbca0d68144c7b317ea8550420ac2594/Python/ceval.c#L2064-L2098

However, this optimization was restricted to only *local* variables within a function, not global/module-scope variables.

Does the regression go away if you put the code in a function?

```python
def f():
    a = ''
    for _ in range(1000000):
        a += 'a'
```

**Comment by zephyr111** [comment]
> Python 3.11 has a specialized (a la [PEP 659](https://peps.python.org/pep-0659/)) instruction called `BINARY_OP_INPLACE_ADD_UNICODE`:
> 
> https://github.com/python/cpython/blob/5bbf8ed8fbca0d68144c7b317ea8550420ac2594/Python/ceval.c#L2064-L2098
> 
> However, this optimization was restricted to only _local_ variables within a function, not global/module-scope variables.
> 
> Does the regression go away if you put the code in a function?

Indeed, this is significantly faster when the code is put in a function: 90 ms in a function in CPython 3.11.0 while it takes 140 ms in CPython 3.10.8.

**Comment by pablogsal** [comment]
```
for _ in range(1000000):
    a += 'a'
end = time.time()
```

that code is quadratic and the only reason is not quadratic in reality is because we do some hacks in the eval loop to reuse the same string, but by the fact that strings are immutable it keeps being quadratic in theory. There is not a lot of reason to spend a lot of resources to speed up this particular code, specially if the regression involves globals. Th correct idiom for this should be:

```
parts = []
for _ in range(100000):
    parts.append('a')
a = "".join(parts)
```

**Comment by eendebakpt** [comment]
Closing this issue is fine with me as well, but I was surprised to find out global/module level code can be slower in 3.11 than local code. Is the behaviour documented somewhere? I could not find it in the [3.11 release notes](https://docs.python.org/3/whatsnew/3.11.html#pep-659-specializing-adaptive-interpreter) or in https://github.com/python/cpython/blob/main/Python/adaptive.md.

**Comment by gvanrossum** [comment]
There are many situations where globals are faster than locals, due to the nature of locals. This is nothing new.

**Comment by markshannon** [comment]
FTR, the linear time behavior will be restored for globals (and nonlocals) in 3.12, as a side-effect of the register VM.

**Comment by markshannon** [comment]
We aren't implementing a register VM, so the performance in 3.12+ will be like 3.11.

Moving the iteration into a function will restore the `n ln(n)` performance.

**Comment by terryjreedy** [comment]
I closed  #115645 as a duplicate of this.

3.10 and whenever before had a hack of internally treated strings like lists by over-allocating space and doing additions in place instead of creating a new string with copies of both strings.  This particularly benefited toy examples of creating a very long string a character at a time.

Trying to sum iterables of strings with `sum` results in `TypeError: sum() can't sum strings [use ''.join(seq) instead]`.  I vaguely remember a FAQ entry.  The [string doc](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) says `There is also no mutable string type, but str.join() or io.StringIO can be used to efficiently construct strings from multiple fragments.`  Perhaps something similar should be added after the discussion of [`+`](https://docs.python.org/3/reference/expressions.html#binary-arithmetic-operations).
